File: next-macos-native/Sources/MooToolNextCore/LegacyCalculatorDraft.swift
import Foundation

enum LegacyCalculatorDraft {
    private static let arithmetic = try! NSRegularExpression(pattern: #"^(.+?)\s*=\s*([^=]+)$"#)
    private static let decFromHex = try! NSRegularExpression(pattern: #"^DEC\(([^)]+)\)\s*=\s*(\d+)$"#, options: [.caseInsensitive])
    private static let hexFromDec = try! NSRegularExpression(pattern: #"^HEX\((\d+)\)\s*=\s*([0-9a-fA-F]+)$"#, options: [.caseInsensitive])
    private static let binFromDec = try! NSRegularExpression(pattern: #"^BIN\((\d+)\)\s*=\s*([01]+)$"#, options: [.caseInsensitive])
    private static let decFromBin = try! NSRegularExpression(pattern: #"^DEC\(([01]+)\)\s*=\s*(\d+)$"#, options: [.caseInsensitive])

    static func apply(to draft: inout DraftRecord, raw: String) {
        let messages = LegacyConsoleDraft.extractMessages(raw)
        let lines = messages.isEmpty
            ? raw.split(whereSeparator: \.isNewline).map { $0.trimmingCharacters(in: .whitespaces) }.filter { !$0.isEmpty }
            : messages
        guard let message = lines.last else { return }
        draft.output = lines.suffix(12).joined(separator: "\n")
        applyMessage(to: &draft, message: message)
    }

    private static func applyMessage(to draft: inout DraftRecord, message: String) {
        let line = message.split(whereSeparator: \.isNewline).map { $0.trimmingCharacters(in: .whitespaces) }.last { !$0.isEmpty } ?? ""
        if line.isEmpty { return }
        let range = NSRange(line.startIndex..., in: line)
        if let match = decFromHex.firstMatch(in: line, range: range) {
            draft.mode = "十六进制 → 十进制"
            draft.input = capture(line, match: match, index: 1).trimmingCharacters(in: .whitespacesAndNewlines)
            draft.secondary = capture(line, match: match, index: 2).trimmingCharacters(in: .whitespacesAndNewlines)
            return
        }
        if let match = hexFromDec.firstMatch(in: line, range: range) {
            draft.mode = "十进制 → 其他进制"
            draft.input = capture(line, match: match, index: 1).trimmingCharacters(in: .whitespacesAndNewlines)
            return
        }
        if let match = binFromDec.firstMatch(in: line, range: range) {
            draft.mode = "十进制 → 其他进制"
            draft.input = capture(line, match: match, index: 1).trimmingCharacters(in: .whitespacesAndNewlines)
            return
        }
        if let match = decFromBin.firstMatch(in: line, range: range) {
            draft.mode = "二进制 → 十进制"
            draft.input = capture(line, match: match, index: 1).trimmingCharacters(in: .whitespacesAndNewlines)
            return
        }
        if let match = arithmetic.firstMatch(in: line, range: range) {
            draft.mode = "表达式"
            draft.input = capture(line, match: match, index: 1).trimmingCharacters(in: .whitespacesAndNewlines)
            draft.secondary = capture(line, match: match, index: 2).trimmingCharacters(in: .whitespacesAndNewlines)
            return
        }
        draft.mode = "表达式"
        draft.input = line
    }

    private static func capture(_ line: String, match: NSTextCheckingResult, index: Int) -> String {
        guard match.numberOfRanges > index, let range = Range(match.range(at: index), in: line) else { return "" }
        return String(line[range])
    }
}

File: next-compose/composeApp/src/desktopMain/kotlin/com/rememberber/mootool/next/compose/domain/LegacyCalculatorDraft.kt
package com.rememberber.mootool.next.compose.domain

import com.rememberber.mootool.next.compose.sessions.CalculatorSession

object LegacyCalculatorDraft {
    private val arithmetic = Regex("""^(.+?)\s*=\s*([^=]+)$""")
    private val decFromHex = Regex("""^DEC\(([^)]+)\)\s*=\s*(\d+)$""", RegexOption.IGNORE_CASE)
    private val hexFromDec = Regex("""^HEX\((\d+)\)\s*=\s*([0-9a-fA-F]+)$""", RegexOption.IGNORE_CASE)
    private val binFromDec = Regex("""^BIN\((\d+)\)\s*=\s*([01]+)$""", RegexOption.IGNORE_CASE)
    private val decFromBin = Regex("""^DEC\(([01]+)\)\s*=\s*(\d+)$""", RegexOption.IGNORE_CASE)

    fun apply(session: CalculatorSession, raw: String) {
        val messages = LegacyConsoleDraft.extractMessages(raw)
        if (messages.isNotEmpty()) {
            session.log = messages.asReversed().take(12)
            applyFromMessage(session, messages.last())
            return
        }
        val lines = raw.lines().map { it.trim() }.filter { it.isNotEmpty() }
        if (lines.isEmpty()) return
        session.log = lines
        applyFromMessage(session, lines.last())
    }

    private fun applyFromMessage(session: CalculatorSession, message: String) {
        val line = message.lineSequence().lastOrNull { it.isNotBlank() }?.trim() ?: return
        decFromHex.matchEntire(line)?.let { match ->
            session.hex = match.groupValues[1].trim()
            session.decimal = match.groupValues[2].trim()
            session.result = match.groupValues[2].trim()
            return
        }
        hexFromDec.matchEntire(line)?.let { match ->
            session.decimal = match.groupValues[1].trim()
            session.hex = match.groupValues[2].trim()
            session.result = match.groupValues[2].trim()
            return
        }
        binFromDec.matchEntire(line)?.let { match ->
            session.decimal = match.groupValues[1].trim()
            session.binary = match.groupValues[2].trim()
            session.result = match.groupValues[2].trim()
            return
        }
        decFromBin.matchEntire(line)?.let { match ->
            session.binary = match.groupValues[1].trim()
            session.decimal = match.groupValues[2].trim()
            session.result = match.groupValues[2].trim()
            return
        }
        arithmetic.matchEntire(line)?.let { match ->
            val expr = match.groupValues[1].trim()
            val value = match.groupValues[2].trim()
            if (expr.isNotEmpty()) session.expression = expr
            if (value.isNotEmpty()) session.result = value
        }
    }
}
